dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/buffer/InjectingPipeWriter.java

```diff
@@ -0,0 +1,160 @@
+package datadog.trace.bootstrap.instrumentation.buffer;
+
+import java.io.IOException;
+import java.io.Writer;
+
+/**
+ * An Writer containing a circular buffer with a lookbehind buffer of n bytes. The first time that
+ * the latest n bytes matches the marker, a content is injected before.
+ */
+public class InjectingPipeWriter extends Writer {
+  private final char[] lookbehind;
+  private int pos;
+  private boolean bufferFilled;
+  private final char[] marker;
+  private final char[] contentToInject;
+  private boolean found = false;
+  private int matchingPos = 0;
+  private final Runnable onContentInjected;
+  private final int bulkWriteThreshold;
+  private final Writer downstream;
+
+  /**
+   * @param downstream the delegate writer
+   * @param marker the marker to find in the stream. Must at least be one char.
+   * @param contentToInject the content to inject once before the marker if found.
+   * @param onContentInjected callback called when and if the content is injected.
+   */
+  public InjectingPipeWriter(
+      final Writer downstream,
+      final char[] marker,
+      final char[] contentToInject,
+      final Runnable onContentInjected) {
+    this.downstream = downstream;
+    this.marker = marker;
+    this.lookbehind = new char[marker.length];
+    this.pos = 0;
+    this.contentToInject = contentToInject;
+    this.onContentInjected = onContentInjected;
+    this.bulkWriteThreshold = marker.length * 2 - 2;
+  }
+
+  @Override
+  public void write(int b) throws IOException {
```

Comment:
```suggestion
  public void write(int c) throws IOException {
```
matches the `Writer` javadoc